keyboards/westberry/common/wb32_driver.c

```diff
@@ -0,0 +1,526 @@
+  tmpcr2   = TIMx->CR2;
+  tmpccmrx = TIMx->CCMR2;
+
+  tmpccmrx &= (uint16_t)(~((uint16_t)TIM_CCMR2_OC4M));
+  tmpccmrx &= (uint16_t)(~((uint16_t)TIM_CCMR2_CC4S));
+
+  tmpccmrx |= (0x70 << 8); // TIM_OCMode_PWM2
+  tmpccer &= (uint16_t)(~((uint16_t)TIM_CCER_CC4P));
+  tmpccer |= (0x00 << 12); // TIM_OCPolarity_High
+  tmpccer |= (0x01 << 12); // TIM_OutputState_Enable
+
+  if (TIMx == TIM1) {
+    tmpccer &= (uint16_t)(~(((uint16_t)0x01) << 12)); // TIM_OutputState_Disable
+    tmpcr2 &= (uint16_t)(~((uint16_t)TIM_CR2_OIS4));
+    tmpcr2 |= (uint16_t)(0x00 << 6); // TIM_OCIdleState_Reset
+  }
+
+  TIMx->CR2   = tmpcr2;
+  TIMx->CCMR2 = tmpccmrx;
+  TIMx->CCR4  = 256;
+  TIMx->CCER  = tmpccer;
+}
+
+static void wb32_pwm_init(void) {
+
+  /* Enable TIM clock */
+  rccEnableTIM1();
+  rccEnableTIM2();
+  rccEnableTIM3();
+  rccEnableTIM4();
+
+  /* Reset TIM */
+  rccResetTIM1();
+  rccResetTIM2();
+  rccResetTIM3();
+  rccResetTIM4();
+
+  /* Config TIM */
+  wb32_tim_base_init(TIM2);
+  wb32_tim_pwm_ch1_init(TIM2);
+  wb32_tim_pwm_ch2_init(TIM2);
+  wb32_tim_pwm_ch3_init(TIM2);
+  wb32_tim_pwm_ch4_init(TIM2);
+  wb32_tim_base_init(TIM3);
+  wb32_tim_pwm_ch1_init(TIM3);
+  wb32_tim_pwm_ch2_init(TIM3);
+  wb32_tim_pwm_ch3_init(TIM3);
+  wb32_tim_pwm_ch4_init(TIM3);
+  wb32_tim_base_init(TIM4);
+  wb32_tim_pwm_ch1_init(TIM4);
+  wb32_tim_pwm_ch2_init(TIM4);
+  wb32_tim_pwm_ch3_init(TIM4);
+  wb32_tim_pwm_ch4_init(TIM4);
+  wb32_tim_base_init(TIM1);
+  wb32_tim_pwm_ch1_init(TIM1);
+  wb32_tim_pwm_ch2_init(TIM1);
+  wb32_tim_pwm_ch3_init(TIM1);
+
+  /* Enable the TIM1 Main Output */
+  TIM1->BDTR |= TIM_BDTR_MOE;
+}
+
+static void wbrgb_init(void) {
+  rgb_driver_init();
+}
+
+static void wbrgb_flush(void) {
+
+  memcpy((void *)g_flush_buffer, (void *)g_pwm_buffer, sizeof(g_flush_buffer));
+}
+
+static void wbrgb_set_color(int index, uint8_t red, uint8_t green, uint8_t blue) {
+  wbrgb_led led;
+
+  memcpy_P(&led, (&g_wbrgb_leds[index]), sizeof(led));
+
+  if (g_pwm_buffer[led.sw][led.id].r != red) {
+    g_pwm_buffer[led.sw][led.id].r = red;
+  }
+  if (g_pwm_buffer[led.sw][led.id].g != green) {
+    g_pwm_buffer[led.sw][led.id].g = green;
+  }
+  if (g_pwm_buffer[led.sw][led.id].b != blue) {
+    g_pwm_buffer[led.sw][led.id].b = blue;
+  }
+}
+
+static void wbrgb_set_color_all(uint8_t red, uint8_t green, uint8_t blue) {
+
+  for (uint8_t i = 0; i < RGB_MATRIX_LED_COUNT; i++) {
+    wbrgb_set_color(i, red, green, blue);
+  }
+}
+
+const rgb_matrix_driver_t rgb_matrix_driver = {
+  .init          = wbrgb_init,
+  .flush         = wbrgb_flush,
+  .set_color     = wbrgb_set_color,
+  .set_color_all = wbrgb_set_color_all
+};
+
+void rgb_driver_init(void) {
+  static bool is_initialised = false;
+
+  if (!is_initialised) {
+    is_initialised = true;
+    wb32_pwm_init();
+    wb32_pwm_pin_init();
+    // Start TIM1.
+    TIM1->CR1 |= TIM_CR1_CEN;
+    wbrgb_set_color_all(0x00, 0x00, 0x00);
+  }
+}
+
+
+/* The following is Key Matrix related. */
```

Comment:
why is there switch matrix and rgb matrix code here?

Reply:
This reply is the same as above, matrix cols are multiplexed with RGB SW Pin.